### ISSUE TITLE
Bleadperl v5.21.0-129-g8dab3ba breaks MLEHMANN/Guard-1.022.tar.gz

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13903]
Migrated from [rt.perl.org#122031](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=122031) (status was 'resolved')

Searchable as RT122031$

**Comment by p5pRT** [comment]
### From @andk

git bisect


---
commit 8dab3ba5fe05d49fe039bc05012fdab8bc572a5c
Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
Date&#8203;:   Tue Apr 8 18&#8203;:46&#8203;:25 2014 \-0600

&nbsp;    Remove deprecated 'PL\_sv\_objcount'

diagnostics


---
cc \-c   \-D\_REENTRANT \-D\_GNU\_SOURCE \-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-O2 \-g   \-DVERSION=\\"1\.022\\" \-DXS\_VERSION=\\"1\.022\\" \-fPIC "\-I/home/sand/src/perl/repoperls/installed\-perls/perl/v5\.21\.0\-129\-g8dab3ba/9980/lib/5\.21\.1/x86\_64\-linux\-thread\-multi/CORE"   Guard\.c
Guard\.xs&#8203;: In function ‘XS\_Guard\_guard’&#8203;:
Guard\.xs&#8203;:113&#8203;:11&#8203;: error&#8203;: ‘PL\_sv\_objcount’ undeclared \(first use in this function\)
&nbsp;         \+\+PL\_sv\_objcount;
&nbsp;           ^
Guard\.xs&#8203;:113&#8203;:11&#8203;: note&#8203;: each undeclared identifier is reported only once for each function it appears in

perl \-V


---
Summary of my perl5 \(revision 5 version 21 subversion 1\) configuration&#8203;:           Commit id&#8203;: 8dab3ba5fe05d49fe039bc05012fdab8bc572a5c                             Platform&#8203;:                                                                         osname=linux\, osvers=3\.13\-1\-amd64\, archname=x86\_64\-linux\-thread\-multi           uname='linux k83 3\.13\-1\-amd64 \#1 smp debian 3\.13\.7\-1 \(2014\-03\-25\) x86\_64 gnulinux '                                                                             config\_args='\-Dprefix=/home/sand/src/perl/repoperls/installed\-perls/perl/v5\.21\.0\-129\-g8dab3ba/9980 \-Dmyhostname=k83 \-Dinstallusrbinperl=n \-Uversiononly \-Dusedevel \-des \-Ui\_db \-Duseithreads \-Uuselongdouble \-DDEBUGGING=\-g'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.8\.2'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8  
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/4\.8/include\-fixed /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
&nbsp;    libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lpthread \-lc \-lgdbm\_compat
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lpthread \-lc
&nbsp;    libc=libc\-2\.18\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.18'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib \-fstack\-protector'


Characteristics of this binary \(from libperl\)&#8203;:
&nbsp;  Compile\-time options&#8203;: HAS\_TIMES MULTIPLICITY PERLIO\_LAYERS
&nbsp;                        PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_HASH\_FUNC\_ONE\_AT\_A\_TIME\_HARD
&nbsp;                        PERL\_IMPLICIT\_CONTEXT PERL\_MALLOC\_WRAP
&nbsp;                        PERL\_NEW\_COPY\_ON\_WRITE PERL\_PRESERVE\_IVUV
&nbsp;                        PERL\_USE\_DEVEL USE\_64\_BIT\_ALL USE\_64\_BIT\_INT
&nbsp;                        USE\_ITHREADS USE\_LARGE\_FILES USE\_LOCALE
&nbsp;                        USE\_LOCALE\_COLLATE USE\_LOCALE\_CTYPE
&nbsp;                        USE\_LOCALE\_NUMERIC USE\_PERLIO USE\_PERL\_ATOF
&nbsp;                        USE\_REENTRANT\_API
&nbsp;  Built under linux
&nbsp;  Compiled at Jun  4 2014 02&#8203;:53&#8203;:12
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.21\.0\-129\-g8dab3ba/9980/lib/site\_perl/5\.21\.1/x86\_64\-linux\-thread\-multi
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.21\.0\-129\-g8dab3ba/9980/lib/site\_perl/5\.21\.1
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.21\.0\-129\-g8dab3ba/9980/lib/5\.21\.1/x86\_64\-linux\-thread\-multi
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.21\.0\-129\-g8dab3ba/9980/lib/5\.21\.1
&nbsp;    \.

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @tsee

On 06/04/2014 03&#8203;:05 AM\, \(Andreas J\. Koenig\) \(via RT\) wrote&#8203;:
> \# New Ticket Created by  \(Andreas J\. Koenig\)
> \# Please include the string&#8203;:  \[perl \#122031\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=122031 >
>
>
> git bisect
> \-\-\-\-\-\-\-\-\-\-
> commit 8dab3ba5fe05d49fe039bc05012fdab8bc572a5c
> Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
> Date&#8203;:   Tue Apr 8 18&#8203;:46&#8203;:25 2014 \-0600
>
>      Remove deprecated 'PL\_sv\_objcount'
>
> diagnostics
> \-\-\-\-\-\-\-\-\-\-\-
> Guard\.xs&#8203;: In function ‘XS\_Guard\_guard’&#8203;:
> Guard\.xs&#8203;:113&#8203;:11&#8203;: error&#8203;: ‘PL\_sv\_objcount’ undeclared \(first use in this function\)
>           \+\+PL\_sv\_objcount;
>             ^
> Guard\.xs&#8203;:113&#8203;:11&#8203;: note&#8203;: each undeclared identifier is reported only once for each function it appears in

The fix for this one is literally just to remove that line of code\. IIRC 
reason for deprecation is that PL\_sv\_objcount was not being used in core 
at all and that it was being kept updated \(somewhat\) as useless busy work\.

\-\-Steffen

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rjbs

I'm happy to pass along any tested patch downstream\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @andk

also affected&#8203;: CREAMYG/KinoSearch\-0\.315\.tar\.gz
\-\- 
andreas

**Comment by p5pRT** [comment]
### From @rurban

On Wed Jun 04 00&#8203;:40&#8203;:25 2014\, smueller@&#8203;cpan\.org wrote&#8203;:
> On 06/04/2014 03&#8203;:05 AM\, \(Andreas J\. Koenig\) \(via RT\) wrote&#8203;:
> > \# New Ticket Created by  \(Andreas J\. Koenig\)
> > \# Please include the string&#8203;:  \[perl \#122031\]
> > \# in the subject line of all future correspondence about this issue\.
> > \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=122031 >
> >
> >
> > git bisect
> > \-\-\-\-\-\-\-\-\-\-
> > commit 8dab3ba5fe05d49fe039bc05012fdab8bc572a5c
> > Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
> > Date&#8203;:   Tue Apr 8 18&#8203;:46&#8203;:25 2014 \-0600
> >
> > Remove deprecated 'PL\_sv\_objcount'
> >
> > diagnostics
> > \-\-\-\-\-\-\-\-\-\-\-
> > Guard\.xs&#8203;: In function ‘XS\_Guard\_guard’&#8203;:
> > Guard\.xs&#8203;:113&#8203;:11&#8203;: error&#8203;: ‘PL\_sv\_objcount’ undeclared \(first use in
> > this function\)
> >           \+\+PL\_sv\_objcount;
> >             ^
> > Guard\.xs&#8203;:113&#8203;:11&#8203;: note&#8203;: each undeclared identifier is reported only
> > once for each function it appears in
> 
> The fix for this one is literally just to remove that line of code\.
> IIRC
> reason for deprecation is that PL\_sv\_objcount was not being used in
> core
> at all and that it was being kept updated \(somewhat\) as useless busy
> work\.

Let's explain it like this&#8203;:
The fix is to check the version\.

<details><summary>Inline Patch</summary>

```diff
--- ./Guard.xs~	2009-07-18 23:58:41.000000000 -0500
+++ ./Guard.xs	2014-06-27 10:18:45.821008691 -0500
@@ -110,7 +110,9 @@
         sv_magicext (guard, cv, PERL_MAGIC_ext, &guard_vtbl, 0, 0);
         RETVAL = newRV_noinc (guard);
         SvOBJECT_on (guard);
+#if PERL_VERSION < 18
         ++PL_sv_objcount;
+#endif
         SvSTASH_set (guard, (HV*)SvREFCNT_inc ((SV *)guard_stash));
 }
 	OUTPUT:



```
</details>
bless does not inc PL_sv_objcount since 5.17.11 anymore and does global destruction without checking for PL_sv_objcount, which slows down global destruction in the case of small scripts without any blessed objects.

5\.21 removed now the variable at all\, so that's why you get this error only now\. 
But it was effectively unused since 5\.18\.

\-\- 
Reini Urban

**Comment by p5pRT** [comment]
### From @andk

Also affected&#8203;:

IBB/Acme\-Damn\-0\.05\.tar\.gz | https://rt.cpan.org/Ticket/Display.html?id=96872

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Jul 01 17&#8203;:49&#8203;:10 2014\, andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de wrote&#8203;:
> Also affected&#8203;:
> 
> IBB/Acme\-Damn\-0\.05\.tar\.gz | https://rt.cpan.org/Ticket/Display.html?id=96872
> 

I have submitted a patch for Acme\-Damn at https://rt.cpan.org/Ticket/Display.html?id=96872.

**Comment by p5pRT** [comment]
### From @karenetheridge

On Fri\, Jun 27\, 2014 at 08&#8203;:25&#8203;:31AM \-0700\, Reini Urban via RT wrote&#8203;:
> Let's explain it like this&#8203;:
> The fix is to check the version\.

Having encountered this problem while trying to install
Test&#8203;::DependentModules \(which requires MetaCPAN&#8203;::Client\, which requires
Search&#8203;::Elasticsearch\, which requires Log&#8203;::Any&#8203;::Adapter\, which requires
Guard\)\, I filed https://rt.cpan.org/Ticket/Display.html?id=96919
with Reini's patch\.

**Comment by p5pRT** [comment]
### From @andk

Perlpolicy says this about backwards incompatibility&#8203;:

&nbsp;    Generally\, backward incompatible changes will have deprecation
&nbsp;    warnings for two release cycles before being removed

It should be noted that the compilation of Guard\.xs did not produce a
warning up to v5\.21\.0\-128\. Admitted\, the citation aims at user\-visible
changes\, but I think that by analogy one could righteously expect that
compilation of XS modules also has a deprecation cycle\. Especially when
there are no pressing needs\.

Are there pressing needs or is it cheap? If the latter\, please consider
rewriting in a backwards compatible way\.

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @tonycoz

On Thu Aug 28 19&#8203;:54&#8203;:52 2014\, andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de wrote&#8203;:
> Perlpolicy says this about backwards incompatibility&#8203;:
> 
>     Generally\, backward incompatible changes will have deprecation
>     warnings for two release cycles before being removed
> 
> It should be noted that the compilation of Guard\.xs did not produce a
> warning up to v5\.21\.0\-128\. Admitted\, the citation aims at user\-visible
> changes\, but I think that by analogy one could righteously expect that
> compilation of XS modules also has a deprecation cycle\. Especially when
> there are no pressing needs\.
> 
> Are there pressing needs or is it cheap? If the latter\, please consider
> rewriting in a backwards compatible way\.

It was documented as deprecated in 5\.18\.0&#8203;:

 =head2 PL\_sv\_objcount

 This interpreter\-global variable used to track the total number of
 Perl objects in the interpreter\. It is no longer maintained and will
 be removed altogether in Perl v5\.20\.

So it did have a deprecation cycle\.

We could look at adding a mechanism for deprecating interpreter and global variables\, but I'm not sure it's worth it for something we've rarely\* done\.

Tony

\*I think

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Jul 01 19&#8203;:02&#8203;:19 2014\, jkeenan wrote&#8203;:
> On Tue Jul 01 17&#8203;:49&#8203;:10 2014\, andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de
> wrote&#8203;:
> > Also affected&#8203;:
> >
> > IBB/Acme\-Damn\-0\.05\.tar\.gz |
> > https://rt.cpan.org/Ticket/Display.html?id=96872
> >
> 
> I have submitted a patch for Acme\-Damn at
> https://rt.cpan.org/Ticket/Display.html?id=96872.

The maintainer of Acme\-Damn has corrected the problem\.  See&#8203;: https://rt.cpan.org/Ticket/Display.html?id=96872

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @rjbs

This has been fixed downstream\.

It would have been nice to have a warning about this during compilation\, but as Tony C\. said\, the cost may have been prohibitive\.

This is yet more evidence that what we need is a revamped set of compatibility guarantees for the XS layer\, which is very much a free\-for\-all where everybody gets to feel badly about any changes made\.
 
\-\- 
rjbs

**Comment by p5pRT** [comment]
@rjbs - Status changed from 'open' to 'resolved'
